v2/tools/mangle-test-json/main.go

```diff
@@ -0,0 +1,224 @@
+			}
+			startTimes[key(d)] = d.Time
+		} else if d.Action == "output" {
+			outputs[key(d)] = append(outputs[key(d)], d.Output)
+		} else if d.Action == "pass" || d.Action == "fail" || d.Action == "skip" {
+			if d.Test != "" && startTimes[key(d)] == (time.Time{}) {
+				panic("finished when not running")
+			}
+
+			runTimes[key(d)] += d.Time.Sub(startTimes[key(d)])
+
+			byPackage[d.Package] = append(byPackage[d.Package], TestRun{
+				Action:  d.Action,
+				Package: d.Package,
+				Test:    d.Test,
+				Output:  outputs[key(d)],
+				RunTime: runTimes[key(d)],
+			})
+		}
+	}
+
+	return byPackage
+}
+
+func printSummary(packages []string, byPackage map[string][]TestRun) {
+	fmt.Printf("## Package Summary\n\n")
+
+	// output table-of-contents
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		if len(tests) == 1 {
+			// assertion: tests[0].Action == "skip"
+			if tests[0].Action != "skip" {
+				panic(fmt.Sprintf("expected only 'skip' for empty package, got: %s", tests[0].Action))
+			}
+			continue
+		}
+
+		overallOutcome := actionSymbol(tests[0])
+		fmt.Printf("* %s `%s`\n", overallOutcome, pkg)
+	}
+
+	fmt.Println()
+}
+
+func printDetails(packages []string, byPackage map[string][]TestRun) {
+	fmt.Printf("## Details\n\n")
+
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		if len(tests) == 1 {
+			// assertion: tests[0].Action == "skip"
+			if tests[0].Action != "skip" {
+				panic(fmt.Sprintf("expected only 'skip' for empty package, got: %s", tests[0].Action))
+			}
+			continue
+		}
+
+		// first test is the "" test for overall package outcome
+		overallOutcome := actionSymbol(tests[0])
+		fmt.Printf("### %s `%s`\n\n", overallOutcome, pkg)
+
+		fmt.Println("| Outcome | Name | Time | Output |")
+		fmt.Println("|---------|------|-----:|--------|")
+		for _, test := range tests[1:] {
+			output := `<details><pre>` + escapeOutput(test.Output) + `</pre></details>`
+			fmt.Printf("| %s | `%s` | %s | %s |\n", actionSymbol(test), test.Test, test.RunTime, output)
+		}
+
+		fmt.Println()
+	}
+}
+
+func escapeOutput(outputs []string) string {
+	result := strings.Builder{}
+	for _, output := range outputs {
+		s := output
+		s = strings.ReplaceAll(s, "&", "&amp;")
+		s = strings.ReplaceAll(s, "<", "&lt;")
+		s = strings.ReplaceAll(s, "\n", "<br>")
+		result.WriteString(s)
+	}
+
+	return result.String()
+}
+
+func printSlowTests(byPackage map[string][]TestRun) {
+	fmt.Printf("## Longest-running tests\n\n")
+
+	allTests := []TestRun{}
+	for _, v := range byPackage {
+		for _, t := range v[1:] { // skip "" package test
+			allTests = append(allTests, t)
+		}
+	}
+
+	sort.Slice(allTests, func(i, j int) bool {
+		return allTests[i].RunTime > allTests[j].RunTime
+	})
+
+	fmt.Println("| Package | Name | Time |")
+	fmt.Println("|---------|------|-----:|")
+	for i := 0; i < min(10, len(allTests)); i += 1 {
+		test := allTests[i]
+		fmt.Printf("| `%s` | `%s` | %s |\n", test.Package, test.Test, test.RunTime)
+	}
```

Comment:
I've got a helper class in the generator for writing a markdown table - see `internal/reporting/markdown-table.go`.

Reply:
I'm not sure it's worth padding the markdown to look nice, the output here is going to be rather unreadable in a text editor. For example, the test output is all in one line and formatted with HTML (this cannot be achieved in “pure markdown”). It's only meant to look good when rendered in the Actions page!